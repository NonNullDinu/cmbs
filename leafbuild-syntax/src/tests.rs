fn parses_without_errors(s: &str) {
    assert_eq!(crate::parser::parse(s).errors, vec![]);
}

// Those are only *parsing* tests, so as long as they make sense
// *syntactically*, they should parse.
// Even if semantically they are void of meaning.
#[test]
fn unit_to_unit_assignment() {
    parses_without_errors("() = ()\n")
}

#[test]
fn line_trailing_space() {
    parses_without_errors("() = () \n")
}

#[test]
fn if_condition() {
    parses_without_errors("if (()) {} else {}")
}

#[test]
fn if_else_if_condition() {
    parses_without_errors("if () {} else if () {} else {}")
}

#[test]
fn if_else_if_else_condition_stretched() {
    parses_without_errors(
        r#"
    if ()

    {



    }

    else if () {


    }

    else {


    }
    "#,
    )
}

#[test]
fn if_condition_in_expr() {
    parses_without_errors("() = if () {} else {} \n")
}

#[test]
fn if_condition_in_expr_stretched() {
    parses_without_errors(
        r#"
    () = if ()
    {

    }

    else
    {


    }
    "#,
    )
}

#[test]
fn declaration() {
    parses_without_errors("let x = ()\n")
}

#[test]
fn foreach_basic() {
    parses_without_errors("foreach () in () {}")
}
