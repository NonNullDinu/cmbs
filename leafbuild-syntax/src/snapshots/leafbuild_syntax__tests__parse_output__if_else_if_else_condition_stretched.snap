---
source: leafbuild-syntax/src/tests.rs
expression: "\n        if ()\n    \n        {\n    \n    \n    \n        }\n    \n        else if () {\n    \n    \n        }\n    \n        else {\n    \n    \n        }\n    "

---
- ROOT 0..140
  - Conditional 0..140
    - ConditionalBranch 0..54
      - "\n" NEWLINE 0..1
      - "        " WHITESPACE 1..9
      - "if" IF_KW 9..11
      - " " WHITESPACE 11..12
      - Expr 12..14
        - PrimaryExpr 12..14
          - TupleExpr 12..14
            - "(" L_PAREN 12..13
            - ")" R_PAREN 13..14
      - ExprBlock 14..54
        - "\n" NEWLINE 14..15
        - "    " WHITESPACE 15..19
        - "\n" NEWLINE 19..20
        - "        " WHITESPACE 20..28
        - "{" L_BRACE 28..29
        - "\n" NEWLINE 29..30
        - "    " WHITESPACE 30..34
        - "\n" NEWLINE 34..35
        - "    " WHITESPACE 35..39
        - "\n" NEWLINE 39..40
        - "    " WHITESPACE 40..44
        - "\n" NEWLINE 44..45
        - "        " WHITESPACE 45..53
        - "}" R_BRACE 53..54
    - "\n" NEWLINE 54..55
    - "    " WHITESPACE 55..59
    - "\n" NEWLINE 59..60
    - "        " WHITESPACE 60..68
    - "else" ELSE_KW 68..72
    - " " WHITESPACE 72..73
    - ConditionalBranch 73..100
      - "if" IF_KW 73..75
      - " " WHITESPACE 75..76
      - Expr 76..78
        - PrimaryExpr 76..78
          - TupleExpr 76..78
            - "(" L_PAREN 76..77
            - ")" R_PAREN 77..78
      - ExprBlock 78..100
        - " " WHITESPACE 78..79
        - "{" L_BRACE 79..80
        - "\n" NEWLINE 80..81
        - "    " WHITESPACE 81..85
        - "\n" NEWLINE 85..86
        - "    " WHITESPACE 86..90
        - "\n" NEWLINE 90..91
        - "        " WHITESPACE 91..99
        - "}" R_BRACE 99..100
    - "\n" NEWLINE 100..101
    - "    " WHITESPACE 101..105
    - "\n" NEWLINE 105..106
    - "        " WHITESPACE 106..114
    - "else" ELSE_KW 114..118
    - " " WHITESPACE 118..119
    - ExprBlock 119..140
      - "{" L_BRACE 119..120
      - "\n" NEWLINE 120..121
      - "    " WHITESPACE 121..125
      - "\n" NEWLINE 125..126
      - "    " WHITESPACE 126..130
      - "\n" NEWLINE 130..131
      - "        " WHITESPACE 131..139
      - "}" R_BRACE 139..140

